rpc: print P2WSH and P2SH redem Script in getrawtransaction and getblock

polespinasa opened the pull request:
This PR is motivated by https://github.com/bitcoin/bitcoin/issues/27391.
And inspired by a previous PR https://github.com/bitcoin/bitcoin/pull/8849 that proposed something similar. 

This is an example with a real mainnet transaction (3-5 multisig) and using ``getrawtransaction 'txid' 2`` and `getblock`. It also affects the endpoint `rest/getblock`

Verbosity must be set to 2 or higher if possible.

```json
$ bitcoin-cli getrawtransaction 44f5ab666cb9dccef69120fe5b51757a647c1f96882931a9faa0b2b5781ddfd2 2 000000000000000000018fa64b5e32335f7d3ac14b9e5553db20f2909c902c34
{
  "in_active_chain": true,
  "txid": "44f5ab666cb9dccef69120fe5b51757a647c1f96882931a9faa0b2b5781ddfd2",
  "hash": "53836d2a4fb1e14356eac69fe1dad8be5dfd9a1b8101bc8aeb04ea2a320aa398",
  "version": 1,
  "size": 382,
  "vsize": 192,
  "weight": 766,
  "locktime": 0,
  "vin": [
    {
      "txid": "d15236206fbe06d6f38505656a2006a9a5a1ddf189a2647750091fd218d59581",
      "vout": 2,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "",
        "304402204d91eb67d798d071278ea0570973f3488d60c3ec98a611856f6a813b2e35df3202200bb36403ebd2dcc96b236efaf07db4f1da26b1031fc854548634ff44a340dc6601",
        "3044022046abaa75082c2d1cdef57f8ef59a0046729220eabe2c2a229d1d592b89ccf1b0022028954703f13d836b1b77acd1eb72639a7a918d756fdd3ddef1d42a28567403ed01",
        "522102f69779f4bb466a8c0aa7d6bf2a2dde34e44bae29d5186ee0c3bed4d903d57fa7210333c337637453164d5fd0dc31787a614a74d2acca5c08710937b9b831bd41b11f2103be7c86e40ad380b33f9abd653c6d04b95b818abd5653b63402678fcc45a017e753ae"
      ],
      "witnessScript": {
        "asm": "2 02f69779f4bb466a8c0aa7d6bf2a2dde34e44bae29d5186ee0c3bed4d903d57fa7 0333c337637453164d5fd0dc31787a614a74d2acca5c08710937b9b831bd41b11f 03be7c86e40ad380b33f9abd653c6d04b95b818abd5653b63402678fcc45a017e7 3 OP_CHECKMULTISIG",
        "desc": "multi(2,02f69779f4bb466a8c0aa7d6bf2a2dde34e44bae29d5186ee0c3bed4d903d57fa7,0333c337637453164d5fd0dc31787a614a74d2acca5c08710937b9b831bd41b11f,03be7c86e40ad380b33f9abd653c6d04b95b818abd5653b63402678fcc45a017e7)#sll5t8hh",
        "type": "multisig"
      },
      "prevout": {
        ... bla bla bla prev output info ...
        }
      },
      "sequence": 4294967293
    }
  ],
  "vout": [ ... bla bla bla outputs ...  ],
  "fee": 0.00000594,
  "hex": "010000000001018195d518d21f09507764a289f1dda1a5a906206a650585f3d606be6f203652d10200000000fdffffff025b730500000000001976a914381d46badae879d1c6037a03da52074e3914812c88ac38088000000000002200205ab691555ede8432ed7765585edb1c9b285532daf639e3e5ccd4843ff3b04cdf040047304402204d91eb67d798d071278ea0570973f3488d60c3ec98a611856f6a813b2e35df3202200bb36403ebd2dcc96b236efaf07db4f1da26b1031fc854548634ff44a340dc6601473044022046abaa75082c2d1cdef57f8ef59a0046729220eabe2c2a229d1d592b89ccf1b0022028954703f13d836b1b77acd1eb72639a7a918d756fdd3ddef1d42a28567403ed0169522102f69779f4bb466a8c0aa7d6bf2a2dde34e44bae29d5186ee0c3bed4d903d57fa7210333c337637453164d5fd0dc31787a614a74d2acca5c08710937b9b831bd41b11f2103be7c86e40ad380b33f9abd653c6d04b95b818abd5653b63402678fcc45a017e753ae00000000",
  "blockhash": "000000000000000000018fa64b5e32335f7d3ac14b9e5553db20f2909c902c34",
  "confirmations": 1,
  "time": 1746727602,
  "blocktime": 1746727602
}

```

polespinasa committed:
rpc: Decode P2(W)SH scripts if data from prev tx is available and verbose is 2

Adds include_scripts (default false) to TxDocOptions to allow specifying if redeemScript and
witnessScript fields must be added to the RPC documentation.

Adds GetRedeemAndWitnessScripts() which takes a CScript object and a CTransaction input and
returns the reedem script and witness script depending if it is a P2WSH, P2SH or P2SH-P2WSH.
Adds a struct RedeemAndWitnessScripts for returning the scripts.

Updates TxToUniv to include redeemScript and witnessScript if the verbose and the undo data allows it,
fields include asm, desc, type and adress.

kevkevinpal commented on src/core_write.cpp:
```diff
@@ -207,6 +207,14 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
+                CScript witnessScript(tx.vin[i].scriptWitness.stack.back().begin(), tx.vin[i].scriptWitness.stack.back().end());
+                UniValue witScript(UniValue::VOBJ);
+                ScriptToUniv(witnessScript, witScript, false, true);
+                in.pushKV("witScript", std::move(witScript));
```
I think it might make more sense to name it "witnessScript" instead of "witScript", It might be confusing to some if it is abbreviated
```suggestion
                in.pushKV("witnessScript", std::move(witScript));
```

polespinasa replied:
I agree, it needs to be changed in one more place apart from this line but yes I will apply it.
Thanks!

kevkevinpal commented:
can we squash 66ea8941de3dfd498319c6390655a165f6a5a92b and 84fb299de72ca9e08fe5029380a779d57d302347
to a single commit

luke-jr commented on src/core_write.cpp:
```diff
@@ -207,6 +207,14 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
             }
             in.pushKV("txinwitness", std::move(txinwitness));
+            CPubKey pubkey(tx.vin[i].scriptWitness.stack.back());
+            if (!pubkey.IsFullyValid()){
```
This isn't an indicator of p2wsh... Maybe use d1f0ff0e1fed2c4124d0d81f18daf7b76a770dda or similar?

polespinasa replied:
code rewritten based on this comment

naiyoma reviewed (commented):
I've tested https://github.com/bitcoin/bitcoin/pull/31252/commits/43dbc696b9e9bb5472b01708d33e022042b5ce3b on regtest, and both (P2SH redeemScript and P2WSH witnessScript )are only being decompiled when calling getblock … 2. I think you should consider editing the PR description to indicate exactly what is working so far.

polespinasa commented:
> I've tested [43dbc69](https://github.com/bitcoin/bitcoin/commit/43dbc696b9e9bb5472b01708d33e022042b5ce3b) on regtest, and both (P2SH redeemScript and P2WSH witnessScript )are only being decompiled when calling getblock … 2. I think you should consider editing the PR description to indicate exactly what is working so far.

Thanks for the review and testing! Yes, the description was outdated to the first implementation of it. Updated 🫡

naiyoma commented on src/core_write.cpp:
```diff
@@ -201,23 +201,31 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
+            UniValue redeemScript(UniValue::VOBJ);
+            ScriptToUniv(script.first, redeemScript, false, true);
+            if (script.second != -2) {
+                in.pushKV("redeemScript", std::move(redeemScript));
```
this outputs `redeemScript` for both P2SH and P2WSH, but it should be 
P2SH `redeemScript` and P2WSH `witnessScript`

polespinasa replied:
done

naiyoma commented on src/core_write.cpp:
```diff
@@ -201,23 +201,31 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
 
             amt_total_in += prev_txout.nValue;
 
+            std::pair<CScript, int> script = GetScriptForTransactionInput(prev_txout.scriptPubKey, txin);
```
This is not a suggestion to replace your current approach, but rather an alternative way. I have an implementation here ->[https://github.com/naiyoma/bitcoin/pull/1/commits/ff82e551404e55c75b1950e01e4f4626f5ad452e](https://github.com/naiyoma/bitcoin/pull/1/commits/ff82e551404e55c75b1950e01e4f4626f5ad452e)  that shows how to decompile  outside of `if(have_undo) `. While it needs more testing, it successfully works for both `getrawtransaction `and `getblock..2`. You might find some ideas there for handling` getrawtransaction `.

polespinasa replied:
Thanks for the suggestion, sorry for the delay, got this PR a bit abandoned, re-taking it now :).

The problem is that I don't see the way to detect when is P2SH or P2WSH without the previous transaction.
Your patch is similar to my first approach. If you test with a P2SH-P2WPKH, the decoding is done wrong.

polespinasa commented:
Rebased on top of master.

naiyoma commented on src/script/script.cpp:
```diff
@@ -387,3 +388,44 @@ bool CheckMinimalPush(const std::vector<unsigned char>& data, opcodetype opcode)
+    if (!prevScript.IsWitnessProgram(witnessversion, witnessprogram)) {
+        // For P2SH, scriptSig is always push-only, so the actual script is only the last stack item
+        // For non-P2SH, prevScript is likely the real script, but not part of this transaction, and scriptSig could very well be executable, so return the latter instead
+        return std::make_pair(p2sh ? prevScript : txin.scriptSig, -1);
```
For a non-P2SH transaction, returning` -1` will result in a `redeemScript` field being added.
I tested this using a P2PKH transaction, and this was the output: `"redeemScript": { "asm": "304402...7981", "type": "nonstandard" }"`

polespinasa replied:
Nice catch! 
You're right, solved in 1d156fce6f1634ca796e4e1ecf1763494f7d4fb2.
Now non-P2SH and non-segwit inputs return -2 and don't show any script as it's out the scope of the PR.

polespinasa committed:
test: test that getrawtransaction with verbosity 2 returns the redeem and witness script

sedited commented on src/script/script.h:
```diff
@@ -573,6 +573,9 @@ class CScript : public CScriptBase
     }
 };
 
+class CTxIn;
```
We usually put forward declarations at the top of the headers.

polespinasa replied:
done

sedited commented on src/core_io.cpp:
```diff
@@ -476,6 +476,15 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
 
             amt_total_in += prev_txout.nValue;
 
+            std::pair<CScript, int> script = GetScriptForTransactionInput(prev_txout.scriptPubKey, txin);
```
`TxToUniv` is used by a few other callers besides the rawtransaction rpcs. Are they effected by this change too?

polespinasa replied:
~Yes, all of them if I am not wrong.~ Will update the PR description.

Update: Only `getblock` is affected and the `/rest/block` endpoint.

`decodepsbt` calls TxToUniv() with the default `txundo=nullptr`.
`gettransaction` explicitly passes `txundo=nullptr`.

So they are not affected as without `txundo` the data is not available.

sedited commented on src/script/script.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <hash.h>
 #include <uint256.h>
 #include <util/hash_type.h>
+#include<script/interpreter.h>
```
I think it would be good to avoid including the interpreter here. Looking at the existing contents of this file, I also don't think this extractor really fits in here. Could we teak script/sign.cpp's `DataFromTransaction` to return the required `SignatureData` from the transactions instead?

polespinasa replied:
>  Could we teak script/sign.cpp's DataFromTransaction to return the required SignatureData from the transactions instead?

Didn't check how to do it but, should we?

```
// Extracts signatures and scripts from incomplete scriptSigs. Please do not extend this, use PSBT instead
SignatureData DataFromTransaction(const CMutableTransaction& tx, unsigned int nIn, const CTxOut& txout)
{
```

polespinasa replied:
I moved the function from script.cpp to solver.cpp I am not sure if it is where it should be, but moved it to avoid a recursive inclusion.

sedited reviewed (commented):
Concept ACK

Is returning the descriptor of the redeem script useful for anything? There have been a few changes in some of the rpc code, can you do a rebase for them? This also still needs a release note.

polespinasa commented:
> Is returning the descriptor of the redeem script useful for anything?

I am not sure, I added it as "free" extra information in case it is useful for something I am not aware of.

Added the field following sjors idea from the original issue :)
> I think adding fields is fine, unless determining their values is slow for some reason.

Jokes a side, I am happy dropping it if we don't want it.

polespinasa committed:
doc: add release notes for redeemScript/witnessScript RPC fields

sedited commented on src/core_io.cpp:
```diff
@@ -473,6 +473,15 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
+                in.pushKV("redeemScript", std::move(dScript));
+            } else if (script.second != -2) {
+                in.pushKV("witnessScript", std::move(dScript));
+            }
```
In the release notes you say that the redeem script is populated for both case, but I think it is either or, no? I think I'd prefer including both, but then I'd also return a struct with optional fields or something instead of status codes that need to be interpreted here.

sedited commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -249,6 +251,77 @@ def getrawtransaction_verbosity_tests(self):
+        # Check that get raw transaction correctly shows the redeem script
+        res_p2sh_tx = node.getrawtransaction(txid=p2sh_tx, verbosity=2, blockhash=blk)
+        assert 'redeemScript' in res_p2sh_tx['vin'][0]
+        assert_equal(['asm', 'desc', 'type'].sort(), list(res_p2sh_tx['vin'][0]['redeemScript'].keys()).sort())
```
Don't you want the following here?
```diff
diff --git a/test/functional/rpc_rawtransaction.py b/test/functional/rpc_rawtransaction.py
index d36cdbd397..6307b786ea 100755
--- a/test/functional/rpc_rawtransaction.py
+++ b/test/functional/rpc_rawtransaction.py
@@ -312 +312 @@ class RawTransactionsTest(BitcoinTestFramework):
-        assert_equal(['asm', 'desc', 'type'].sort(), list(res_p2sh_tx['vin'][0]['redeemScript'].keys()).sort())
+        assert_equal(sorted(['asm', 'desc', 'type']), sorted(res_p2sh_tx['vin'][0]['redeemScript'].keys()))
@@ -320 +320 @@ class RawTransactionsTest(BitcoinTestFramework):
-        assert_equal(['asm', 'desc', 'type'].sort(), list(res_p2wsh_tx['vin'][0]['witnessScript'].keys()).sort())
+        assert_equal(sorted(['asm', 'desc', 'type']), sorted(res_p2wsh_tx['vin'][0]['witnessScript'].keys()))
```

polespinasa replied:
taken

sedited commented on src/core_io.cpp:
```diff
@@ -473,6 +473,15 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
 
+            std::pair<CScript, int> script = GetScriptForTransactionInput(prev_txout.scriptPubKey, txin);
+            UniValue dScript(UniValue::VOBJ);
+            ScriptToUniv(script.first, dScript, false, true);
```
Can you add argument doc hints to all the arguments here? If this includes the address shouldn't that be in the docs too and get a functional test exercising the field?

polespinasa committed:
refactor: make use of already created variables for better readability

polespinasa committed:
test: test getblock with verbosity returns redeemscript and witness script if possible

polespinasa commented:
I refactored a bit the PR and added some better testing.
I think now the code is clearer and it is easier to review :)

Also took some changes suggested by sedited.
